version : '3.8'
networks:
  my_network:
    driver: bridge
services:
  zhao-nacos:
    container_name: zhao-nacos
    restart: always
    image: nacos/nacos-server
    build:
      context: ./nacos
    environment:
      - MODE=standalone
    volumes:
      - ./nacos/logs/:/home/nacos/logs
      - ./nacos/conf/application.properties:/home/nacos/conf/application.properties
    ports:
      - "8848:8848"
      - "9848:9848"
      - "9849:9849"
    depends_on:
      - zhao-mysql
    networks:
      - my_network
    healthcheck:
      test: ["CMD", "curl", "-fL", "http://zhao-nacos:8848/nacos"]
      interval: 10s
      timeout: 5s
      retries: 3
  zhao-mysql:
    container_name: zhao-mysql
    image: mysql:5.7
    networks:
      - my_network
    build:
      context: ./mysql
    ports:
      - "3306:3306"
    volumes:
      - ./mysql/conf:/etc/mysql/conf.d
      - ./mysql/logs:/logs
      - ./mysql/data:/var/lib/mysql
    command: [
      'mysqld',
      '--innodb-buffer-pool-size=80M',
      '--character-set-server=utf8mb4',
      '--collation-server=utf8mb4_unicode_ci',
      '--default-time-zone=+8:00',
      '--lower-case-table-names=1'
    ]
    environment:
      MYSQL_DATABASE: 'zhao-cloud'
      MYSQL_ROOT_PASSWORD: '123456'
  zhao-redis:
    container_name: zhao-redis
    image: redis
    networks:
      - my_network
    build:
      context: ./redis
    ports:
      - "6379:6379"
    volumes:
      - ./redis/conf/redis.conf:/home/zhao/redis/redis.conf
      - ./redis/data:/data
    command: redis-server /home/zhao/redis/redis.conf
    healthcheck: # 添加健康检测
      test: [ "CMD", "redis-cli", "ping" ]
      interval: 30s
      timeout: 3s
      retries: 3
      start_period: 40s
  zhao-ldap:
    container_name: zhao-ldap
    restart: always
    networks:
      - my_network
    build:
      context: ./ldap
      dockerfile: dockerfile
    ports:
      - "389:389"
      - "636:636"
    command:
      - "--copy-service"
      - "--loglevel=debug"
    healthcheck:
      test: [ "CMD", "ldapsearch", "-x", "-H", "ldap://zhao-ldap:389", "-D", "cn=admin,dc=example,dc=org", "-w", "admin", "-b", "dc=example,dc=org" ]
      interval: 10s
      timeout: 5s
      retries: 3
  zhao-nginx:
    container_name: zhao-nginx
    restart: always
    image: nginx
    networks:
      - my_network
    build:
      context: ./nginx
    ports:
      - "80:80"
    volumes:
      - ./nginx/html/dist:/home/zhao/projects/zhao-ui
      - ./nginx/conf/nginx.conf:/etc/nginx/nginx.conf
      - ./nginx/logs:/var/log/nginx
      - ./nginx/conf.d:/etc/nginx/conf.d
    depends_on:
      - zhao-gateway
  zhao-gateway:
    container_name: zhao-gateway
    restart: always
    networks:
      - my_network
    environment:
      NACOS_HOST: zhao-nacos
    build:
      context: ./zhao/gateway
      dockerfile: dockerfile
    ports:
      - "7573:7573"
    depends_on:
      zhao-redis:
        condition: service_healthy
      zhao-nacos:
        condition: service_healthy
  zhao-auth:
    container_name: zhao-auth
    restart: always
    networks:
      - my_network
    environment:
      NACOS_HOST: zhao-nacos
    build:
      context: ./zhao/auth
      dockerfile: dockerfile
    ports:
      - "9200:9200"
    depends_on:
      zhao-redis:
        condition: service_healthy
      zhao-nacos:
        condition: service_healthy
      zhao-ldap:
        condition: service_healthy
  zhao-modules-system:
    container_name: zhao-modules-system
    restart: always
    networks:
      - my_network
    environment:
      NACOS_HOST: zhao-nacos
    build:
      context: ./zhao/modules/system
      dockerfile: dockerfile
    ports:
      - "9201:9201"
    depends_on:
      zhao-redis:
        condition: service_healthy
      zhao-mysql:
        condition: service_started
      zhao-nacos:
        condition: service_healthy
